### PR TITLE
JAVA-1761: Add OSGi descriptors, upgrade to Guava 25.1-jre

### PR DESCRIPTION
**Pull request by tolbertam**
Motivation:

To enable OSGi support in the DataStax Java drivers, this module must
also support OSGi.

Modifications:

Use maven-bundle-plugin to add OSGi information to the generated
MANIFEST.MF.  Because maven-bundle-plugin runs before
maven-shade-plugin, we also explicitly provide the Export-Package and
Import-Package values, which match the guava jar's values.

Also upgrades to Guava 25.1-jre which was released in May.  Since the
version needs to be increased anyways, this seemed like a good
opportunity to upgrade.

Result:

This module now supports OSGi, upgrades to Guava 25.1-jre.

Generated MANIFEST.MF for reference:

```
Manifest-Version: 1.0                                                                                                                                                                                                                                                                                                                              
Bnd-LastModified: 1528216286603
Build-Jdk: 1.8.0_172
Built-By: andrew.tolbert
Bundle-Description: Shaded Guava artifact for use in the DataStax Java
  driver for Apache Cassandra®
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0.txt
Bundle-ManifestVersion: 2
Bundle-Name: Shaded Guava artifact for use in the DataStax Java driver
  for Apache Cassandra®
Bundle-SymbolicName: com.datastax.oss.driver.shaded.guava
Bundle-Version: 25.1.0.jre-SNAPSHOT
Created-By: Apache Maven Bundle Plugin
Export-Package: com.datastax.oss.driver.shaded.guava.common.annotation
 s;version="25.1.0",com.datastax.oss.driver.shaded.guava.common.base;v
 ersion="25.1.0",com.datastax.oss.driver.shaded.guava.common.cache;ver
 sion="25.1.0";uses:="com.datastax.oss.driver.shaded.guava.common.base
 ,com.datastax.oss.driver.shaded.guava.common.collect,com.datastax.oss
 .driver.shaded.guava.common.util.concurrent",com.datastax.oss.driver.
 shaded.guava.common.collect;version="25.1.0";uses:="com.datastax.oss.
 driver.shaded.guava.common.base",com.datastax.oss.driver.shaded.guava
 .common.escape;version="25.1.0";uses:="com.datastax.oss.driver.shaded
 .guava.common.base",com.datastax.oss.driver.shaded.guava.common.event
 bus;version="25.1.0",com.datastax.oss.driver.shaded.guava.common.grap
 h;version="25.1.0";uses:="com.datastax.oss.driver.shaded.guava.common
 .collect",com.datastax.oss.driver.shaded.guava.common.hash;version="2
 5.1.0";uses:="com.datastax.oss.driver.shaded.guava.common.base",com.d
 atastax.oss.driver.shaded.guava.common.html;version="25.1.0";uses:="c
 om.datastax.oss.driver.shaded.guava.common.escape",com.datastax.oss.d
 river.shaded.guava.common.io;version="25.1.0";uses:="com.datastax.oss
 .driver.shaded.guava.common.base,com.datastax.oss.driver.shaded.guava
 .common.collect,com.datastax.oss.driver.shaded.guava.common.graph,com
 .datastax.oss.driver.shaded.guava.common.hash",com.datastax.oss.drive
 r.shaded.guava.common.math;version="25.1.0",com.datastax.oss.driver.s
 haded.guava.common.net;version="25.1.0";uses:="com.datastax.oss.drive
 r.shaded.guava.common.base,com.datastax.oss.driver.shaded.guava.commo
 n.collect,com.datastax.oss.driver.shaded.guava.common.escape",com.dat
 astax.oss.driver.shaded.guava.common.primitives;version="25.1.0";uses
 :="com.datastax.oss.driver.shaded.guava.common.base",com.datastax.oss
 .driver.shaded.guava.common.reflect;version="25.1.0";uses:="com.datas
 tax.oss.driver.shaded.guava.common.collect,com.datastax.oss.driver.sh
 aded.guava.common.io",com.datastax.oss.driver.shaded.guava.common.uti
 l.concurrent;version="25.1.0";uses:="com.datastax.oss.driver.shaded.g
 uava.common.base,com.datastax.oss.driver.shaded.guava.common.collect"
 ,com.datastax.oss.driver.shaded.guava.common.xml;version="25.1.0";use
 s:="com.datastax.oss.driver.shaded.guava.common.escape"
Import-Package: javax.annotation;resolution:=optional;version="[3.0,4)
 ",javax.crypto;resolution:=optional,javax.crypto.spec;resolution:=opt
 ional,sun.misc;resolution:=optional
Tool: Bnd-3.5.0.201709291849
```

Guava 25.1-jre MANIFEST.MF:

```
Manifest-Version: 1.0                                                                                                                                                                                   
Automatic-Module-Name: com.google.common
Bnd-LastModified: 1527098197718
Build-Jdk: 1.8.0_131-google-v7
Built-By: cgdecker
Bundle-Description: Guava is a suite of core and expanded libraries that
  include    utility classes, google's collections, io classes, and much
     much more.
Bundle-DocURL: https://github.com/google/guava/
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0.txt
Bundle-ManifestVersion: 2
Bundle-Name: Guava: Google Core Libraries for Java
Bundle-SymbolicName: com.google.guava
Bundle-Version: 25.1.0.jre
Created-By: Apache Maven Bundle Plugin
Export-Package: com.google.common.annotations;version="25.1.0",com.googl
 e.common.base;version="25.1.0",com.google.common.cache;version="25.1.0"
 ;uses:="com.google.common.base,com.google.common.collect,com.google.com
 mon.util.concurrent",com.google.common.collect;version="25.1.0";uses:="
 com.google.common.base",com.google.common.escape;version="25.1.0";uses:
 ="com.google.common.base",com.google.common.eventbus;version="25.1.0",c
 om.google.common.graph;version="25.1.0";uses:="com.google.common.collec
 t",com.google.common.hash;version="25.1.0";uses:="com.google.common.bas
 e",com.google.common.html;version="25.1.0";uses:="com.google.common.esc
 ape",com.google.common.io;version="25.1.0";uses:="com.google.common.bas
 e,com.google.common.collect,com.google.common.graph,com.google.common.h
 ash",com.google.common.math;version="25.1.0",com.google.common.net;vers
 ion="25.1.0";uses:="com.google.common.base,com.google.common.collect,co
 m.google.common.escape",com.google.common.primitives;version="25.1.0";u
 ses:="com.google.common.base",com.google.common.reflect;version="25.1.0
 ";uses:="com.google.common.collect,com.google.common.io",com.google.com
 mon.util.concurrent;version="25.1.0";uses:="com.google.common.base,com.
 google.common.collect",com.google.common.xml;version="25.1.0";uses:="co
 m.google.common.escape"
Import-Package: javax.annotation;resolution:=optional;version="[3.0,4)",
 javax.crypto;resolution:=optional,javax.crypto.spec;resolution:=optiona
 l,sun.misc;resolution:=optional
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-2.3.0.201405100607
```

### CODE REVIEW

**Comment by tolbertam** [comment]
Lets keep this open until we have everything worked out with OSGi with driver 4.x

File: pom.xml [hunk 1]
```diff
@@ -99,6 +106,65 @@
...
+        <groupId>org.apache.felix</groupId>
+        <artifactId>maven-bundle-plugin</artifactId>
+        <version>3.5.0</version>
+        <extensions>true</extensions>
+        <configuration>
+          <instructions>
+            <Bundle-SymbolicName>com.datastax.oss.driver.shaded.guava</Bundle-SymbolicName>
+            <!-- Export-Package and Import-Package must be declared explicitly since the maven-bundle-plugin executes before maven-shade-plugin, so the guava classes aren't present at this time.
+                 Note that these should match the specifications from the MANIFEST.mf file of the guava jar this jar will be based on. -->
+            <Export-Package>
```
Comment (tolbertam) [inline pom.xml]: This does look kind of hacky, but was best solution I could find.  Also it's pretty easy to extract from the Guava MANIFEST and then do a search and replace.

File: pom.xml [hunk 2]
```diff
@@ -99,6 +106,65 @@
...
         </configuration>
       </plugin>
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>build-helper-maven-plugin</artifactId>
+        <version>3.0.0</version>
+        <!-- To match the OSGi version format in the Export-Package specification of the manifest, strip off
+             the 'jre' part of the version name (and anything following it, i.e. SNAPSHOT). -->
+        <executions>
+          <execution>
```
Comment (tolbertam) [inline pom.xml]: OSGi is very particular about versions 😬

**Comment by tolbertam** [comment]
I removed the whole `${osgi.version}` business as it wasn't needed.  Originally was having issues because I was not quoting the version number.  It also looks like the bundle plugin properly sanitizes `${project.version}` in the MANIFEST even when specified manually (i.e `25.1-jre-SNAPSHOT` becomes `25.1.0.jre-SNAPSHOT`:

```
Manifest-Version: 1.0                                                                                                                                                                                                                                                                                                                         
Bnd-LastModified: 1528387023041
Build-Jdk: 1.8.0_172
Built-By: andrew.tolbert
Bundle-Description: Shaded Guava artifact for use in the DataStax Java
  driver for Apache Cassandra®
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0.txt
Bundle-ManifestVersion: 2
Bundle-Name: Shaded Guava artifact for use in the DataStax Java driver
  for Apache Cassandra®
Bundle-SymbolicName: com.datastax.oss.driver.shaded.guava
Bundle-Version: 25.1.0.jre-SNAPSHOT
Created-By: Apache Maven Bundle Plugin
Export-Package: com.datastax.oss.driver.shaded.guava.common.annotation
 s;version="25.1.0.jre-SNAPSHOT",com.datastax.oss.driver.shaded.guava.
 common.base;version="25.1.0.jre-SNAPSHOT",com.datastax.oss.driver.sha
 ded.guava.common.cache;version="25.1.0.jre-SNAPSHOT";uses:="com.datas
 tax.oss.driver.shaded.guava.common.base,com.datastax.oss.driver.shade
 d.guava.common.collect,com.datastax.oss.driver.shaded.guava.common.ut
 il.concurrent",com.datastax.oss.driver.shaded.guava.common.collect;ve
 rsion="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.driver.shaded.gua
 va.common.base",com.datastax.oss.driver.shaded.guava.common.escape;ve
 rsion="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.driver.shaded.gua
 va.common.base",com.datastax.oss.driver.shaded.guava.common.eventbus;
 version="25.1.0.jre-SNAPSHOT",com.datastax.oss.driver.shaded.guava.co
 mmon.graph;version="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.driv
 er.shaded.guava.common.collect",com.datastax.oss.driver.shaded.guava.
 common.hash;version="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.dri
 ver.shaded.guava.common.base",com.datastax.oss.driver.shaded.guava.co
 mmon.html;version="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.drive
 r.shaded.guava.common.escape",com.datastax.oss.driver.shaded.guava.co
 mmon.io;version="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.driver.
 shaded.guava.common.base,com.datastax.oss.driver.shaded.guava.common.
 collect,com.datastax.oss.driver.shaded.guava.common.graph,com.datasta
 x.oss.driver.shaded.guava.common.hash",com.datastax.oss.driver.shaded
 .guava.common.math;version="25.1.0.jre-SNAPSHOT",com.datastax.oss.dri
 ver.shaded.guava.common.net;version="25.1.0.jre-SNAPSHOT";uses:="com.
 datastax.oss.driver.shaded.guava.common.base,com.datastax.oss.driver.
 shaded.guava.common.collect,com.datastax.oss.driver.shaded.guava.comm
 on.escape",com.datastax.oss.driver.shaded.guava.common.primitives;ver
 sion="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.driver.shaded.guav
 a.common.base",com.datastax.oss.driver.shaded.guava.common.reflect;ve
 rsion="25.1.0.jre-SNAPSHOT";uses:="com.datastax.oss.driver.shaded.gua
 va.common.collect,com.datastax.oss.driver.shaded.guava.common.io",com
 .datastax.oss.driver.shaded.guava.common.util.concurrent;version="25.
 1.0.jre-SNAPSHOT";uses:="com.datastax.oss.driver.shaded.guava.common.
 base,com.datastax.oss.driver.shaded.guava.common.collect",com.datasta
 x.oss.driver.shaded.guava.common.xml;version="25.1.0.jre-SNAPSHOT";us
 es:="com.datastax.oss.driver.shaded.guava.common.escape"
Import-Package: javax.annotation;resolution:=optional;version="[3.0,4)
 ",javax.crypto;resolution:=optional,javax.crypto.spec;resolution:=opt
 ional,sun.misc;resolution:=optional
Tool: Bnd-3.5.0.201709291849
```

File: pom.xml (hunk 1 above)
Comment (olim7t) [reply pom.xml]: I tried fixing the execution order but it's not enough.

If you use non-default executions for both plugins, they will run in the order they're declared in the POM:

```xml
      <plugin>
        <artifactId>maven-shade-plugin</artifactId>
        <executions>
          <execution>
            <id>relocate-classes</id>
            <goals><goal>shade</goal></goals>
            <configuration> ... </configuration>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <artifactId>maven-bundle-plugin</artifactId>
        <executions>
          <execution>
            <id>write-osgi-manifest</id>
            <goals><goal>bundle</goal></goals>
            <configuration> ... </configuration>
          </execution>
        </executions>
      </plugin>
```
The shade plugin replaces the artifact first:
```
[INFO] Replacing original artifact with shaded artifact.
```
But the bundle plugin still doesn't see it:

```
[WARNING] Bundle com.datastax.oss:java-driver-shaded-guava:bundle:25.1-jre-SNAPSHOT : The JAR is empty: The instructions for the JAR named com.datastax.oss.driver.shaded.guava did not cause any content to be included, this is likely wrong
```

I ran into similar issues while working on shading the driver. It looks like the JAR is replaced at the filesystem level, but other plugins in the reactor are still operating on the pre-shaded structure (maybe held in memory).

Anyway this module shouldn't change often, so it's not worth spending too much time on this.

**Comment by tolbertam** [comment]
> On a side note, I hadn't anticipated the need to release new versions for anything else than a Guava upgrade, so it's a good thing a new version was available :/ If this becomes a problem in the future, maybe we should add a 3rd component to the version number.

Yeah I was trying to figure out what the best route to go was in terms of bumping the version, but fortunately there was a new Guava release 😄.  I agree that we should wait until we run into this issue before we come up with a solution.  Fortunately Guava seems to release about twice a year now, and I doubt we'll need to change this as much (if it all), so 🤞 that we'll never have to worry about it.

**Comment by tolbertam** [comment]
Will merge after driver CI run (just to make sure my commit reorg didn't break anything)

**Comment by tolbertam** [comment]
CI ran well, merging and proceeding with release.

**Comment by olim7t** [comment]
:shipit:

**Pull request merged**